third_party/NordicSemiconductor/libraries/crypto/aes_alt.c

```diff
@@ -0,0 +1,151 @@
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+/*
+ *  Copyright (C) 2006-2015, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "mbedtls/aes.h"
+
+#include <openthread-core-config.h>
+#include <string.h>
+
+#ifdef MBEDTLS_AES_ALT
+
+#include "aes_alt_cc310.h"
+
+#if OPENTHREAD_CONFIG_INTERRUPT_CONTEXT_AES
+#include "aes_alt_soft.h"
+#endif
+
+#if OPENTHREAD_CONFIG_INTERRUPT_CONTEXT_AES
+void mbedtls_aes_set_interrupt_context(mbedtls_aes_context * ctx, bool interrupt)
+{
+    ctx->interrupt_context = interrupt;
+}
+#endif
+
+void mbedtls_aes_init(mbedtls_aes_context * ctx)
+{
+#if OPENTHREAD_CONFIG_INTERRUPT_CONTEXT_AES
+    if (!ctx->interrupt_context)
+    {
+#endif
+        aes_cc310_init(ctx);
+#if OPENTHREAD_CONFIG_INTERRUPT_CONTEXT_AES
+        ctx->using_cc310 = true;
+    }
+    else
+    {
+        aes_soft_init(ctx);
```

Comment:
Do we always want interrupt context to use software-based implementation?  If the hardware peripheral is available, should the interrupt handler utilize that?

Reply:
Yes, it's as designed for nRF52840 platform, since the hardware Aes (based on cc310) could not support interrupt context.

The node will hang when doing AES in interrupt, which seems an interrupt nesting issue:
```
(gdb) bt
#0  0x00035a0a in SaSi_HalWaitInterrupt ()
#1  0x00035c20 in ProcessAes ()
```

For other platforms which could support hardware Aes in interrupt context, yes, we should prefer hardware peripheral when it is available.

Reply:
Thanks for pointing that out.  Given that this limitation seems platform specific, is there a better way to determine at runtime whether execution is occurring within interrupt context or not?  That would eliminate any need for OT core to provide that signal to the platform layer and simplify the overall implementation.

Reply:
I have updated the PR to do interrupt context check at runtime, it does help to simplify the OT core implementation, thanks for the suggestion.

I have verified it with #2618, it works well. Please help to review again, thanks.

Reply:
Great, thanks!